src/realm/object-store/sync/sync_session.cpp

```diff
@@ -1536,3 +1595,29 @@ util::Future<std::string> SyncSession::send_test_command(std::string body)
 
     return m_session->send_test_command(std::move(body));
 }
+
+void SyncSession::make_active_subscription_set()
+{
+    util::CheckedUniqueLock lock(m_state_mutex);
+
+    if (!m_active_subscriptions_after_migration)
+        return;
+
+    REALM_ASSERT(m_flx_subscription_store);
+
+    // Create subscription set from the subscriptions used to download the fresh realm after migration.
+    auto active_mut_sub = m_flx_subscription_store->get_active().make_mutable_copy();
+    active_mut_sub.import(*m_active_subscriptions_after_migration);
+    lock.unlock();
+
+    // Commit without holding the lock to prevent a deadlock.
```

Comment:
Just curious what kind of deadlock you could end up in here?

Reply:
committing a subscription set fires `on_new_subscription_set` which locks m_state_mutex

Reply:
There is no need to release the lock at all anymore, because we can update the state and commit. This prevents on_new_subscription_set to fire (the root of the deadlock)